File: src/arithmetic_fractions_easy.py
"""
arithmetic - fractions (easy)
Generated: 2026-02-22T03:40:17.437521
"""

from problem_utils import *

def generate():
    problem_type = randint(1, 6)
    
    if problem_type == 1:
        # Simplify a fraction (1000-1100)
        numerator = randint(2, 12)
        common_factor = randint(2, 6)
        denominator = numerator * common_factor
        numerator = numerator * common_factor
        
        # Ensure we get a reducible fraction
        from math import gcd
        g = gcd(numerator, denominator)
        simplified_num = numerator // g
        simplified_den = denominator // g
        
        ans = Rational(simplified_num, simplified_den)
        
        return problem(
            question=f"Simplify the fraction: $\\frac{{{numerator}}}{{{denominator}}}$",
            answer=ans,
            difficulty=(1000, 1100),
            topic="arithmetic/fractions",
            solution=steps(
                f"Find the greatest common divisor of ${numerator}$ and ${denominator}$: GCD = ${g}$",
                f"Divide both numerator and denominator by ${g}$",
                f"$\\frac{{{numerator}}}{{{denominator}}} = \\frac{{{numerator} \\div {g}}}{{{denominator} \\div {g}}} = \\frac{{{simplified_num}}}{{{simplified_den}}}$"
            ),
        )
    
    elif problem_type == 2:
        # Add two fractions with same denominator (1000-1100)
        denominator = randint(3, 12)
        num1 = randint(1, denominator - 1)
        num2 = randint(1, denominator - 1)
        
        ans = Rational(num1 + num2, denominator)
        
        return problem(
            question=f"Add the fractions: $\\frac{{{num1}}}{{{denominator}}} + \\frac{{{num2}}}{{{denominator}}}$",
            answer=ans,
            difficulty=(1000, 1100),
            topic="arithmetic/fractions",
            solution=steps(
                f"Since the denominators are the same, add the numerators",
                f"$\\frac{{{num1}}}{{{denominator}}} + \\frac{{{num2}}}{{{denominator}}} = \\frac{{{num1} + {num2}}}{{{denominator}}} = \\frac{{{num1 + num2}}}{{{denominator}}}$",
                f"Simplified: ${latex(ans)}$"
            ),
        )
    
    elif problem_type == 3:
        # Subtract two fractions with same denominator (1100-1200)
        denominator = randint(4, 12)
        num1 = randint(3, denominator)
        num2 = randint(1, num1 - 1)
        
        ans = Rational(num1 - num2, denominator)
        
        return problem(
            question=f"Subtract the fractions: $\\frac{{{num1}}}{{{denominator}}} - \\frac{{{num2}}}{{{denominator}}}$",
            answer=ans,
            difficulty=(1100, 1200),
            topic="arithmetic/fractions",
            solution=steps(
                f"Since the denominators are the same, subtract the numerators",
                f"$\\frac{{{num1}}}{{{denominator}}} - \\frac{{{num2}}}{{{denominator}}} = \\frac{{{num1} - {num2}}}{{{denominator}}} = \\frac{{{num1 - num2}}}{{{denominator}}}$",
                f"Simplified: ${latex(ans)}$"
            ),
        )
    
    elif problem_type == 4:
        # Multiply two simple fractions (1100-1200)
        num1 = randint(1, 6)
        den1 = randint(2, 8)
        num2 = randint(1, 6)
        den2 = randint(2, 8)
        
        ans = Rational(num1 * num2, den1 * den2)
        
        return problem(
            question=f"Multiply the fractions: $\\frac{{{num1}}}{{{den1}}} \\times \\frac{{{num2}}}{{{den2}}}$",
            answer=ans,
            difficulty=(1100, 1200),
            topic="arithmetic/fractions",
            solution=steps(
                f"Multiply numerators and denominators",
                f"$\\frac{{{num1}}}{{{den1}}} \\times \\frac{{{num2}}}{{{den2}}} = \\frac{{{num1} \\times {num2}}}{{{den1} \\times {den2}}} = \\frac{{{num1 * num2}}}{{{den1 * den2}}}$",
                f"Simplified: ${latex(ans)}$"
            ),
        )
    
    elif problem_type == 5:
        # Add fractions with different denominators (1200-1300)
        den1 = randint(2, 6)
        den2 = randint(2, 6)
        while den2 == den1:
            den2 = randint(2, 6)
        
        num1 = randint(1, den1)
        num2 = randint(1, den2)
        
        from math import gcd
        lcm = (den1 * den2) // gcd(den1, den2)
        
        ans = Rational(num1, den1) + Rational(num2, den2)
        
        return problem(
            question=f"Add the fractions: $\\frac{{{num1}}}{{{den1}}} + \\frac{{{num2}}}{{{den2}}}$",
            answer=ans,
            difficulty=(1200, 1300),
            topic="arithmetic/fractions",
            solution=steps(
                f"Find the least common denominator: LCD = ${lcm}$",
                f"Convert to equivalent fractions: $\\frac{{{num1 * (lcm // den1)}}}{{{lcm}}} + \\frac{{{num2 * (lcm // den2)}}}{{{lcm}}}$",
                f"Add: $\\frac{{{num1 * (lcm // den1) + num2 * (lcm // den2)}}}{{{lcm}}}$",
                f"Simplified: ${latex(ans)}$"
            ),
        )
    
    else:
        # Divide two simple fractions (1200-1300)
        num1 = randint(1, 6)
        den1 = randint(2, 8)
        num2 = randint(1, 6)
        den2 = randint(2, 8)
        
        ans = Rational(num1 * den2, den1 * num2)
        
        return problem(
            question=f"Divide the fractions: $\\frac{{{num1}}}{{{den1}}} \\div \\frac{{{num2}}}{{{den2}}}$",
            answer=ans,
            difficulty=(1200, 1300),
            topic="arithmetic/fractions",
            solution=steps(
                f"Multiply by the reciprocal of the second fraction",
                f"$\\frac{{{num1}}}{{{den1}}} \\times \\frac{{{den2}}}{{{num2}}} = \\frac{{{num1} \\times {den2}}}{{{den1} \\times {num2}}} = \\frac{{{num1 * den2}}}{{{den1 * num2}}}$",
                f"Simplified: ${latex(ans)}$"
            ),
        )

emit(generate())

File: src/arithmetic_fractions_easy.jl
# arithmetic - fractions (easy)
# Generated: 2026-03-08T19:47:54.349067

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script begin
    set_topic!("arithmetic/fractions")
    
    prob_type = rand(1:6)
    
    if prob_type == 1
        # Addition of fractions with different denominators (700-900)
        d1 = randint(2, 12)
        d2 = randint(2, 12)
        while gcd(d1, d2) == d1 || gcd(d1, d2) == d2
            d2 = randint(2, 12)
        end
        n1 = randint(1, d1-1)
        n2 = randint(1, d2-1)
        
        lcm_d = lcm(d1, d2)
        result_n = n1 * (lcm_d ÷ d1) + n2 * (lcm_d ÷ d2)
        result_d = lcm_d
        g = gcd(result_n, result_d)
        final_n = result_n ÷ g
        final_d = result_d ÷ g
        
        problem(
            question="Simplify: \$\\frac{$n1}{$d1} + \\frac{$n2}{$d2}\$",
            answer="$final_n/$final_d",
            difficulty=(700, 900),
            solution=steps(
                "Find common denominator: LCM($d1, $d2) = $lcm_d",
                "\$\\frac{$n1}{$d1} = \\frac{$(n1*(lcm_d÷d1))}{$lcm_d}\$ and \$\\frac{$n2}{$d2} = \\frac{$(n2*(lcm_d÷d2))}{$lcm_d}\$",
                "\$\\frac{$(n1*(lcm_d÷d1))}{$lcm_d} + \\frac{$(n2*(lcm_d÷d2))}{$lcm_d} = \\frac{$result_n}{$result_d}\$",
                sol("Simplify", "$final_n/$final_d")
            ),
            time=90
        )
        
    elseif prob_type == 2
        # Subtraction of fractions (700-900)
        d1 = randint(3, 12)
        d2 = randint(3, 12)
        while gcd(d1, d2) == d1 || gcd(d1, d2) == d2
            d2 = randint(3, 12)
        end
        n1 = randint(d1 ÷ 2, d1-1)
        n2 = randint(1, d2 ÷ 2)
        
        lcm_d = lcm(d1, d2)
        result_n = n1 * (lcm_d ÷ d1) - n2 * (lcm_d ÷ d2)
        result_d = lcm_d
        g = gcd(abs(result_n), result_d)
        final_n = result_n ÷ g
        final_d = result_d ÷ g
        
        problem(
            question="Simplify: \$\\frac{$n1}{$d1} - \\frac{$n2}{$d2}\$",
            answer="$final_n/$final_d",
            difficulty=(700, 900),
            solution=steps(
                "Find common denominator: LCM($d1, $d2) = $lcm_d",
                "\$\\frac{$n1}{$d1} = \\frac{$(n1*(lcm_d÷d1))}{$lcm_d}\$ and \$\\frac{$n2}{$d2} = \\frac{$(n2*(lcm_d÷d2))}{$lcm_d}\$",
                "\$\\frac{$(n1*(lcm_d÷d1))}{$lcm_d} - \\frac{$(n2*(lcm_d÷d2))}{$lcm_d} = \\frac{$result_n}{$result_d}\$",
                sol("Simplify", "$final_n/$final_d")
            ),
            time=90
        )
        
    elseif prob_type == 3
        # Multiplication of fractions (800-1000)
        n1 = randint(2, 15)
        d1 = randint(2, 15)
        n2 = randint(2, 15)
        d2 = randint(2, 15)
        
        result_n = n1 * n2
        result_d = d1 * d2
        g = gcd(result_n, result_d)
        final_n = result_n ÷ g
        final_d = result_d ÷ g
        
        problem(
            question="Multiply and simplify: \$\\frac{$n1}{$d1} \\times \\frac{$n2}{$d2}\$",
            answer="$final_n/$final_d",
            difficulty=(800, 1000),
            solution=steps(
                "Multiply numerators and denominators",
                "\$\\frac{$n1 \\times $n2}{$d1 \\times $d2} = \\frac{$result_n}{$result_d}\$",
                "GCD($result_n, $result_d) = $g",
                sol("Simplify", "$final_n/$final_d")
            ),
            time=75
        )
        
    elseif prob_type == 4
        # Division of fractions (800-1000)
        n1 = randint(2, 12)
        d1 = randint(2, 12)
        n2 = randint(2, 12)
        d2 = randint(2, 12)
        
        result_n = n1 * d2
        result_d = d1 * n2
        g = gcd(result_n, result_d)
        final_n = result_n ÷ g
        final_d = result_d ÷ g
        
        problem(
            question="Divide and simplify: \$\\frac{$n1}{$d1} \\div \\frac{$n2}{$d2}\$",
            answer="$final_n/$final_d",
            difficulty=(800, 1000),
            solution=steps(
                "Multiply by the reciprocal: \$\\frac{$n1}{$d1} \\times \\frac{$d2}{$n2}\$",
                "\$\\frac{$n1 \\times $d2}{$d1 \\times $n2} = \\frac{$result_n}{$result_d}\$",
                "GCD($result_n, $result_d) = $g",
                sol("Simplify", "$final_n/$final_d")
            ),
            time=75
        )
        
    elseif prob_type == 5
        # Complex fraction addition/subtraction (900-1100)
        d1 = randint(4, 10)
        d2 = randint(4, 10)
        d3 = randint(4, 10)
        n1 = randint(1, d1-1)
        n2 = randint(1, d2-1)
        n3 = randint(1, d3-1)
        
        op = rand(["+", "-"])
        
        lcm_d = lcm(lcm(d1, d2), d3)
        if op == "+"
            result_n = n1 * (lcm_d ÷ d1) + n2 * (lcm_d ÷ d2) + n3 * (lcm_d ÷ d3)
        else
            result_n = n1 * (lcm_d ÷ d1) + n2 * (lcm_d ÷ d2) - n3 * (lcm_d ÷ d3)
        end
        result_d = lcm_d
        g = gcd(abs(result_n), result_d)
        final_n = result_n ÷ g
        final_d = result_d ÷ g
        
        problem(
            question="Simplify: \$\\frac{$n1}{$d1} + \\frac{$n2}{$d2} $op \\frac{$n3}{$d3}\$",
            answer="$final_n/$final_d",
            difficulty=(900, 1100),
            solution=steps(
                "Find common denominator: LCM($d1, $d2, $d3) = $lcm_d",
                "Convert each fraction: \$\\frac{$(n1*(lcm_d÷d1))}{$lcm_d} + \\frac{$(n2*(lcm_d÷d2))}{$lcm_d} $op \\frac{$(n3*(lcm_d÷d3))}{$lcm_d}\$",
                "Combine: \$\\frac{$result_n}{$result_d}\$",
                sol("Simplify", "$final_n/$final_d")
            ),
            time=120
        )
        
    else
        # Mixed operations (1000-1200)
        n1 = randint(2, 10)
        d1 = randint(2, 10)
        n2 = randint(2, 10)
        d2 = randint(2, 10)
        n3 = randint(2, 10)
        d3 = randint(2, 10)
        
        # (n1/d1 * n2/d2) + n3/d3
        mult_n = n1 * n2
        mult_d = d1 * d2
        
        lcm_d = lcm(mult_d, d3)
        result_n = mult_n * (lcm_d ÷ mult_d) + n3 * (lcm_d ÷ d3)
        result_d = lcm_d
        g = gcd(result_n, result_d)
        final_n = result_n ÷ g
        final_d = result_d ÷ g
        
        problem(
            question="Simplify: \$\\left(\\frac{$n1}{$d1} \\times \\frac{$n2}{$d2}\\right) + \\frac{$n3}{$d3}\$",
            answer="$final_n/$final_d",
            difficulty=(1000, 1200),
            solution=steps(
                "First multiply: \$\\frac{$n1}{$d1} \\times \\frac{$n2}{$d2} = \\frac{$mult_n}{$mult_d}\$",
                "Find common denominator with \$\\frac{$n3}{$d3}\$: LCM($mult_d, $d3) = $lcm_d",
                "Add: \$\\frac{$(mult_n*(lcm_d÷mult_d))}{$lcm_d} + \\frac{$(n3*(lcm_d÷d3))}{$lcm_d} = \\frac{$result_n}{$result_d}\$",
                sol("Simplify", "$final_n/$final_d")
            ),
            time=150
        )
    end
end